src/main/cpp/profiler/spark_rapids_profile_converter.cpp

```diff
@@ -0,0 +1,756 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* A tool that converts a spark-rapids profile binary into other forms. */
+
+#if 0
+#include <stdexcept>
+#define FLATBUFFERS_ASSERT(x)                                     \
+  do {                                                            \
+    if (!(x)) { throw std::runtime_error("flatbuffers assert"); } \
+  } while (0)
+#define FLATBUFFERS_DEBUG_VERIFICATION_FAILURE
+#endif
+
+#include "profiler_generated.h"
+#include "spark_rapids_jni_version.h"
+
+#include <cupti.h>
+#include <cxxabi.h>
+#include <flatbuffers/idl.h>
+
+#include <cerrno>
+#include <charconv>
+#include <filesystem>
+#include <fstream>
+#include <iostream>
+#include <optional>
+#include <sstream>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+namespace spark_rapids_jni::profiler {
+extern char const* Profiler_Schema;
+}
+
+struct program_options {
+  std::optional<std::filesystem::path> output_path;
+  bool help       = false;
+  bool json       = false;
+  bool nvtxt      = false;
+  int json_indent = 2;
+  bool version    = false;
+};
+
+struct event {
+  enum struct type_id { API, DEVICE, KERNEL, MARKER, MARKER_DATA, MEMCPY, MEMSET, OVERHEAD };
+  type_id id;
+  void const* fb_data;
+};
+
+struct thread_id {
+  uint32_t pid;
+  uint32_t tid;
+
+  bool operator==(thread_id const& o) const { return pid == o.pid && tid == o.tid; }
+};
+
+template <>
+struct std::hash<thread_id> {
+  std::size_t operator()(thread_id const& t) const
+  {
+    // TODO: use a real hash
```

Comment:
Do we need this? Should there be a follow on issue?

Reply:
I suspect we don't really need to do this in practice, as the hash is probably "good enough" for the limited data we're using with it.